File: library/src/main/java/cn/wandersnail/listview/swipe/SwipeMenuAdapter.java
package cn.wandersnail.listview.swipe;

import android.database.DataSetObserver;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListAdapter;
import android.widget.WrapperListAdapter;

/**
 * date: 2019/8/24 12:25
 * author: zengfansheng
 */
abstract class SwipeMenuAdapter implements WrapperListAdapter, SwipeMenuView.OnSwipeItemClickListener {
    private final ListAdapter adapter;

    SwipeMenuAdapter(ListAdapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public ListAdapter getWrappedAdapter() {
        return adapter;
    }

    @Override
    public boolean areAllItemsEnabled() {
        return adapter.areAllItemsEnabled();
    }

    @Override
    public boolean isEnabled(int position) {
        return adapter.isEnabled(position);
    }

    @Override
    public void registerDataSetObserver(DataSetObserver observer) {
        adapter.registerDataSetObserver(observer);
    }

    @Override
    public void unregisterDataSetObserver(DataSetObserver observer) {
        adapter.unregisterDataSetObserver(observer);
    }

    @Override
    public int getCount() {
        return adapter.getCount();
    }

    @Override
    public Object getItem(int position) {
        return adapter.getItem(position);
    }

    @Override
    public long getItemId(int position) {
        return adapter.getItemId(position);
    }

    @Override
    public boolean hasStableIds() {
        return adapter.hasStableIds();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        SwipeMenuLayout layout;
        SwipeMenuListView lv = (SwipeMenuListView) parent;
        if (convertView == null) {
            View contentView = adapter.getView(position, null, lv);
            SwipeMenu menu = new SwipeMenu();
            menu.itemViewType = getItemViewType(position);
            createMenu(menu);
            SwipeMenuView menuView = new SwipeMenuView(lv.getContext());
            menuView.initialize(menu);
            menuView.listener = this;
            layout = new SwipeMenuLayout(lv.getContext());
            layout.initialize(contentView, menuView, lv.openInterpolator, lv.closeInterpolator);
        } else {
            layout = (SwipeMenuLayout) convertView;
            layout.closeMenu();
            adapter.getView(position, layout.contentView, parent);
        }       
        layout.setPosition(position);
        if (adapter instanceof SwipeController) {
            layout.swipeEnabled = ((SwipeController) adapter).isSwipeEnabled(position);
        }
        return layout;
    }

    @Override
    public int getItemViewType(int position) {
        return adapter.getItemViewType(position);
    }

    @Override
    public int getViewTypeCount() {
        return adapter.getViewTypeCount();
    }

    @Override
    public boolean isEmpty() {
        return adapter.isEmpty();
    }
        
    abstract void createMenu(SwipeMenu menu);
}

File: library/src/main/java/com/snail/swipemenulistview/SwipeMenuAdapter.kt
package com.snail.swipemenulistview

import android.content.Context
import android.database.DataSetObserver
import android.view.View
import android.view.ViewGroup
import android.widget.ListAdapter
import android.widget.WrapperListAdapter

open class SwipeMenuAdapter(private val mContext: Context, private val mAdapter: ListAdapter) : WrapperListAdapter, SwipeMenuView.OnSwipeItemClickListener {
    private var onMenuItemClickListener: SwipeMenuListView.OnMenuItemClickListener? = null

    override fun getCount(): Int {
        return mAdapter.count
    }

    override fun getItem(position: Int): Any {
        return mAdapter.getItem(position)
    }

    override fun getItemId(position: Int): Long {
        return mAdapter.getItemId(position)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val layout: SwipeMenuLayout
        if (convertView == null) {
            val contentView = mAdapter.getView(position, null, parent)
            val menu = SwipeMenu(mContext)
            menu.viewType = getItemViewType(position)
            createMenu(menu)
            val menuView = SwipeMenuView(menu, parent as SwipeMenuListView)
            menuView.onSwipeItemClickListener = this
            layout = SwipeMenuLayout(contentView, menuView, parent.closeInterpolator, parent.openInterpolator)
            layout.position = position
        } else {
            layout = convertView as SwipeMenuLayout
            layout.closeMenu()
            layout.position = position
            mAdapter.getView(position, layout.contentView, parent)
        }
        if (mAdapter is SwipeController) {
            val swipEnable = (mAdapter as SwipeController).getSwipeEnableByPosition(position)
            layout.swipEnable = swipEnable
        }
        return layout
    }

    open fun createMenu(menu: SwipeMenu) {}

    override fun onItemClick(view: SwipeMenuView, menu: SwipeMenu, index: Int) {
        onMenuItemClickListener?.onMenuItemClick(view.position, menu, index)
    }

    fun setOnSwipeItemClickListener(onMenuItemClickListener: SwipeMenuListView.OnMenuItemClickListener?) {
        this.onMenuItemClickListener = onMenuItemClickListener
    }

    override fun registerDataSetObserver(observer: DataSetObserver) {
        mAdapter.registerDataSetObserver(observer)
    }

    override fun unregisterDataSetObserver(observer: DataSetObserver) {
        mAdapter.unregisterDataSetObserver(observer)
    }

    override fun areAllItemsEnabled(): Boolean {
        return mAdapter.areAllItemsEnabled()
    }

    override fun isEnabled(position: Int): Boolean {
        return mAdapter.isEnabled(position)
    }

    override fun hasStableIds(): Boolean {
        return mAdapter.hasStableIds()
    }

    override fun getItemViewType(position: Int): Int {
        return mAdapter.getItemViewType(position)
    }

    override fun getViewTypeCount(): Int {
        return mAdapter.viewTypeCount
    }

    override fun isEmpty(): Boolean {
        return mAdapter.isEmpty
    }

    override fun getWrappedAdapter(): ListAdapter {
        return mAdapter
    }

}
